src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py

```diff
@@ -328,36 +331,56 @@ def cleanRSE(self, rse):
-
-                        # Now clean the whole branch
-                        self.logger.debug("Purging dirEntry: %s:\n", dirPfn)
-                        purgeSuccess = self._purgeTree(ctx, dirPfn)
-                        if not purgeSuccess:
-                            msg = "RSE: %s Failed to purge nonEmpty directory: %s"
-                            self.logger.error(msg, rse['name'], dirPfn)
-                    except Exception as ex:
-                        msg = "Error while cleaning RSE: %s. "
-                        msg += "Will retry in the next cycle. Err: %s"
-                        self.logger.warning(msg, rse['name'], str(ex))
+                            self.logger.debug("RSE: %s, Dir: %s, delResult: %s",
+                                              rse['name'], dirLfn, pformat(delResult))
+                            for gfalErr in delResult:
+                                if gfalErr is None:
+                                    deletedSuccess += 1
+                                elif os.strerror(gfalErr.code) not in rse['counters']['gfalErrors']:
+                                    rse['counters']['gfalErrors'][os.strerror(gfalErr.code)] = 1
+                                else:
+                                    rse['counters']['gfalErrors'][os.strerror(gfalErr.code)] += 1
+                        except Exception as ex:
+                            msg = "Error while cleaning RSE: %s. "
+                            msg += "Will retry in the next cycle. Err: %s"
+                            self.logger.warning(msg, rse['name'], str(ex))
+
+                    self.logger.debug("RSE: %s, Dir: %s, filesDeletedSuccess: %s",
+                                      rse['name'], dirLfn, deletedSuccess)
+                    rse['counters']['filesDeletedSuccess'] += deletedSuccess
+
+                    # Now clean the whole branch
+                    self.logger.debug("Purging dirEntry: %s:\n", dirPfn)
+                    purgeSuccess = self._purgeTree(ctx, dirPfn)
+                    if purgeSuccess:
+                        rse['dirs']['deletedSuccess'].add(dirLfn)
+                        rse['counters']['dirsDeletedSuccess'] = len(rse['dirs']['deletedSuccess'])
+                        if dirLfn in rse['dirs']['toDelete']:
+                            rse['dirs']['toDelete'].remove(dirLfn)
+                        if dirLfn in rse['dirs']['deletedFail']:
+                            rse['dirs']['deletedFail'].remove(dirLfn)
+                        msg = "RSE: %s  Success deleting directory: %s"
+                        self.logger.info(msg, rse['name'], dirPfn)
+                    else:
+                        rse['dirs']['deletedFail'].add(dirLfn)
+                        rse['counters']['dirsDeletedFail'] = len(rse['dirs']['deletedFail'])
+                        msg = "RSE: %s Failed to purge directory: %s"
+                        self.logger.error(msg, rse['name'], dirPfn)
```

Comment:
Do we know why it failed to be purged? What is `purgeSuccess` when it fails? If it's just a boolean, then there is nothing that we can do indeed.

Reply:
it is just a `bool` indeed. The reason for failing is just logged into the exceptions handling during the process. If we were to develop a different behavior driven by the cause of failure, I admit the code would be definitely more precise, but also quite more complicate, because with `gfal` and all those many protocols to be used in the background it may become quite a lot to handle. I would rather leave it for `gfal` to do its excellent job in that regard, while we just count the final successful operations, and use the log messages for debugging the failures.